### PR TITLE
Online dispatcher of actions for RDDL Planners

### PR DESCRIPTION
**Pull request by gerardcanal**
Currently supporting PROST.

### CODE REVIEW

File: rosplan_planning_system/include/rosplan_planning_system/PlanDispatch/EsterelPlanDispatcher.h [hunk 1]
```diff
@@ -58,27 +56,24 @@ namespace KCL_rosplan
...
-		/* check preconditions are true */
-		bool checkPreconditions(rosplan_dispatch_msgs::ActionDispatch msg);
 		ros::ServiceClient query_knowledge_client;
 		ros::ServiceClient query_domain_client;
 
 	public:
 
 		/* constructor */
-		EsterelPlanDispatcher(ros::NodeHandle& nh);
+		explicit EsterelPlanDispatcher(ros::NodeHandle& nh);
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/include/rosplan_planning_system/PlanDispatch/EsterelPlanDispatcher.h]: I think that's my IDE's, I must have accepted some of the suggestions for that file. It just avoids that implicit conversions of the parameters are performed. Can be removed I guess.

File: rosplan_planning_system/include/rosplan_planning_system/PlanDispatch/PlanDispatcher.h [hunk 2]
```diff
@@ -1,20 +1,33 @@
 /**
  * This file describes a class that dispatches a plan.
  */
+
+
+#ifndef KCL_dispatcher
+#define KCL_dispatcher
+
+#include <rosplan_dispatch_msgs/CompletePlan.h>
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/include/rosplan_planning_system/PlanDispatch/PlanDispatcher.h]: The OnlineDispatcher also uses it (keeps adding actions to the CompletePlan when they're executed).

File: rosplan_planning_system/src/PlanDispatch/OnlinePlanDispatcher.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    /*--------------------*/
+    /* Dispatch interface */
+    /*--------------------*/
+
+    /**
+     * plan dispatch service method (1)
+     * dispatches plan as a service
+     * @returns True iff every action was dispatched and returned success.
+     */
+    bool OnlinePlanDispatcher::dispatchPlanService(rosplan_dispatch_msgs::DispatchService::Request &req, rosplan_dispatch_msgs::DispatchService::Response &res) {
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/OnlinePlanDispatcher.cpp]: This is just to set the mission start time, as it is not overriden by the parent call. In the other dispatchers mission_start_time is set in the planner callback. In the case of the online_dispatcher there's no such callback, so I am setting the mission_start_time to the start of the dispatcher (which is when the planning also starts)

File: rosplan_planning_system/src/PlanDispatch/OnlinePlanDispatcher.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        mission_start_time = ros::WallTime::now().toSec();
+        PlanDispatcher::dispatchPlanService(req, res);
+    }
+
+    /**
+     * plan dispatch action method (2)
+     * dispatches plan as a service
+     * @returns True iff every action was dispatched and returned success.
+     */
+    void OnlinePlanDispatcher::dispatchPlanActionlib() {
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/OnlinePlanDispatcher.cpp]: Same reason as above.

File: rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        // Action topics
+        action_dispatch_topic = "action_dispatch";
+        action_feedback_topic = "action_feedback";
+        nh.getParam("action_dispatch_topic", action_dispatch_topic);
+        nh.getParam("action_feedback_topic", action_feedback_topic);
+        action_dispatch_publisher = node_handle->advertise<rosplan_dispatch_msgs::ActionDispatch>(action_dispatch_topic, 1, true);
+        action_feedback_publisher = node_handle->advertise<rosplan_dispatch_msgs::ActionFeedback>(action_feedback_topic, 1, true);
+    }
+
+    void PlanDispatcher::publishFeedback(const rosplan_dispatch_msgs::ActionFeedback &fb) {
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp]: I'm not sure I understand what's the problem here, this action_feedback is the dispatcher one, not the one from actionlib. I agree that the naming is a bit confusing, but it's difficult to change that now as many nodes subscribe to the action_feedback topic.

[the remark this answers is not in the document]
Comment (m312z) [reply rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp]: I don't think it's a problem. When we start adding the status topics and address the feedback email then we should see if either there's a way to clear up this confusion, or at least keep it in mind.

File: rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            rosplan_dispatch_msgs::NonBlockingDispatchFeedback afeedback;
+            afeedback.feedback = fb;
+            as_.publishFeedback(afeedback);
+        }
+        // Publish feedback in topic anyway...
+        action_feedback_publisher.publish(fb);
+    }
+
+    bool PlanDispatcher::cancelDispatch() {
+        if (as_.isNewGoalAvailable()) { // Preempt caused by sending of new goal, ignore it
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp]: Fixed the first comment.
For the second: cancelDispatch is also used as a preempt callback for actionlib, so when you cancel the actionlib action or send a new goal it gets canceled.  As there's no point in starting a dispatch when the dispatch is being performed, the if checks the cause of the cancel. So if there's a new goal waiting to be executed while already dispatching one, it justs ignores that.

File: rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            ROS_ERROR("KCL: (%s) could not call Knowledge Base for operator details, %s", ros::this_node::getName().c_str(), msg.name.c_str());
+            return false;
+        }
+
+        // setup service call
+        rosplan_knowledge_msgs::DomainOperator op = srv.response.op;
+        rosplan_knowledge_msgs::KnowledgeQueryService querySrv;
+
+        // iterate through conditions
+        std::vector<rosplan_knowledge_msgs::DomainFormula>::iterator cit = op.at_start_simple_condition.begin();
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp]: I think I didn't change that, just moved to the upper class. Agree that this should be added, can you assign it to someone? Or we can do it quickly next week together.

File: rosplan_planning_system/src/PlanDispatch/SimplePlanDispatcher.cpp [hunk 8]
```diff
@@ -1,3 +1,6 @@
+
+#include <rosplan_planning_system/PlanDispatch/SimplePlanDispatcher.h>
+
 #include "rosplan_planning_system/PlanDispatch/SimplePlanDispatcher.h"
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/SimplePlanDispatcher.cpp]: That's Clion, I'm tired of deleting double includes, this one made it to the commit. Deleted.

File: rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (m312z) [reply rosplan_planning_system/src/PlanDispatch/PlanDispatcher.cpp]: Yes, next week. I've added it to a small tasks list that we can auction in the next ROSPlan meeting.

File: rosplan_planning_system/src/PlanDispatch/SimplePlanDispatcher.cpp [hunk 9]
```diff
@@ -111,21 +82,33 @@ namespace KCL_rosplan {
...
 				fb.action_id = currentMessage.action_id;
 				fb.status = "precondition false";
-				action_feedback_publisher.publish(fb);
+				publishFeedback(fb);
 
 				replan_requested = true;
 
 			} else {
 
+			    std:: string params = "(";
```
[the remark this answers is not in the document]
Comment (gerardcanal) [reply rosplan_planning_system/src/PlanDispatch/SimplePlanDispatcher.cpp]: Fixed.

**Pull request merged**